### DISCUSSION TITLE
Feedback and ideas

### DISCUSSION THREAD

**Discussion by rich-mond** [pdevito3/craftsman#38] [Ideas]
Hey @pdevito3 cool project, looking forward to 0.12!

I wanted to mention a few things, I am not sure where in GitHub they would be most appropriate to put so ive added them all here:

Potential bugs/issues 0.11.1

These have (yaml definition) workarounds so may not be considered high priority

1.  If entity has the DeleteRecord feature and the singular and plural names are the same it breaks the associated integration test. Variables in the test are both the same name
2.  An entity without defined features creates a empty controller, it contains non existent namespace reference (maybe if no features, don't create controller)
3.  Adding an Authentication section does not add the plumbing to AddInfrastructure, it requires information to be added to Environments

Future issue

Updating NuGet FluentAssertions package to 6.1.0 breaks tests

- In the Functional Tests, there are changes to the syntax of the HttpCodes (from int to constant) 
- In the Integration Tests, exception assert (changed to Async)

Suggestions/queries

1. Will the base entity allow definition of common properties? Eg, lastmodified, isdeleted, etc
2. Will foreign key support provide support for one to one/many relationships?
3. Could certain yaml properties be applied to the whole entity without attaching them to each property? Eg CanSort/CanFilter

**Comment by pdevito3** [comment]
hey, @rich-mond, thanks for the detailed note, this is great!

* For the 11.1 bugs, i made issues for 1 and 2, but could submit an issue for 3, ideally with a template to use so i can see what you mean?
* Good call on Fluent Assertions. I'll take a look at it and make a ticket to cover it.

> Will the base entity allow definition of common properties? Eg, lastmodified, isdeleted, etc

I actually have something called `BaseEntitiy` exactly getting used in 12.x that will be added with a PK by default. probably going to integrate some auth stuff and maybe even soft delete like you mentioned.

> Will foreign key support provide support for one to one/many relationships?

the FK feature currently in 12.x is focused on 1:1, but i'm hoping to add something for 1:many as well.

> Could certain yaml properties be applied to the whole entity without attaching them to each property? Eg CanSort/CanFilter

this is possible, but i don't have plans for it immediately. open to a PR here if you want to take a crack at it

**Comment by rich-mond** [comment]
Hi @pdevito3 

Glad the feedback was useful!

I'll expand on issue number 3.

The first example file contains 2 entities and an AuthorizationSettings section specifying a policy:
[example.yaml.txt](https://github.com/pdevito3/craftsman/files/7149182/example.yaml.txt)

The second file is the same, with the addition of auth properties on the Qa environment:
[exampleWithEnvironment.yaml.txt](https://github.com/pdevito3/craftsman/files/7149183/exampleWithEnvironment.yaml.txt)

AuthorizationSettings decorate the controller methods with the Authorization tag

`        [Authorize(Policy = "Admin")]`

What I think is missing from the first file example generated output (and is present in the second file) is the following code: 

InfrastructureServiceExtension.cs

`            services.AddAuthorization(options =>
            {
                options.AddPolicy("Admin",
                    policy => policy.RequireClaim("scope", "API.Admin"));
            });`

Which is the code that sets the Policy used by the Authorize tags if the correct claim is present. 

I would say if there is AuthorizationSettings with policy specified, then the policy/claim mapping should be there in InfrastructureServiceExtension.cs.

**Comment by pdevito3** [comment]
ah, i see. so, technically this is working as expected, but i should very probably change it in 0.12.x. 

if you're curious, under the hood, there is a boolean property on the boundary object that is checking to see if there are auth settings in `Environment` (specifically if an authority is given) and then use it to scaffold out auth in several areas.

```csharp
        public bool AddJwtAuthentication
        {
            get => Environments
                .Where(e => e?.Authority?.Length > 0)
                .ToList()
                .Count > 0;
        }
```

this probably makes sense in some areas, but others like `Policies` should probably be determined separately, so good catch!
